tools/Launcher/Environment.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE ApplicativeDo         #-}
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf            #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE QuasiQuotes           #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeSynonymInstances  #-}
+
+module Launcher.Environment
+  (substituteEnvVars)
+where
+
+import qualified Prelude (show)
+import           Universum
+
+-- * Environment variable parsing and substitution for the launcher configuration file,
+--   typically launcher-config.yaml.
+
+-- | Represent an element of a parse of a string containing environment variable
+--   references: either a plain text piece, or a variable reference.
+data Chunk
+  = EnvVar Text
+  | Plain  Text
+  deriving (Show)
+
+-- | A parse is a sequence of variable references & plain text pieces.
+parseEnvrefs :: Text -> P.Result [Chunk]
+parseEnvrefs text = P.parseString pEnvrefs mempty (toString text)
+  where
+    pEnvrefs :: (Monad p, P.TokenParsing p) => p [Chunk]
+    pEnvrefs = P.some $ P.choice [pPassiveText, pEnvvarRef]
+
+    pEnvvarRef, pPassiveText :: (Monad p, P.TokenParsing p) => p Chunk
+    -- TODO: figure out a less ugly OS conditionalisation method
+    -- TODO: decide if we want escaping
```

Comment:
Must decide on this before merging. Did we previously support escaping?

Reply:
We didn't no.